ranger/config/commands.py

```diff
@@ -1176,24 +1177,86 @@ def execute(self):
 
 
 class chmod(Command):
-    """:chmod <octal number>
+    """:chmod [<octal number>|<mode-spec>]
 
-    Sets the permissions of the selection to the octal number.
+    Sets the permissions of the selection to the given mode.
 
     The octal number is between 0 and 777. The digits specify the
     permissions for the user, the group and others.
 
     A 1 permits execution, a 2 permits writing, a 4 permits reading.
     Add those numbers to combine them. So a 7 permits everything.
+
+    The <mode-spec> is a comma-separated where each entry has the
+    form [ugoa]+[+-][rwx]+
     """
 
-    def execute(self):
-        mode_str = self.rest(1)
+    def __init__(self, *args, **kwargs):
+        '''init chmod-command.'''
+        super(chmod, self).__init__(*args, **kwargs)
+        self.re_mode = None
+        self.bitmasks = None
+
+    def prepare(self):
+        '''create regex for mode-spec and bitmasks.'''
+        self.re_mode = re.compile(r'''
+            ([ugoa]+)     # (u)ser, (g)roup, (o)other, (a)ll
+            ([+-])        # add (+), remove (-)
+            ([rwx]+)      # (r)ead, (w)rite, e(x)ecute
+            $             # end
+            ''', re.X)
+        masks = {
+            'u': {'r': stat.S_IRUSR, 'w': stat.S_IWUSR, 'x': stat.S_IXUSR},
+            'g': {'r': stat.S_IRGRP, 'w': stat.S_IWGRP, 'x': stat.S_IXGRP},
+            'o': {'r': stat.S_IROTH, 'w': stat.S_IWOTH, 'x': stat.S_IXOTH}}
+        masks['a'] = {
+            'r': masks['u']['r'] | masks['g']['r'] | masks['o']['r'],
+            'w': masks['u']['w'] | masks['g']['w'] | masks['o']['w'],
+            'x': masks['u']['x'] | masks['g']['x'] | masks['o']['x']}
+        self.bitmasks = masks
+
+    def get_mask_for(self, ugoa_part, rwx_part):
```

Comment:
Let's go with the names from the chmod grammar "who" and "perm".

Reply:
We can use [this man page as a reference](https://www.man7.org/linux/man-pages/man1/chmod.1p.html) for the grammar. (Or the BSD chmod man page because that's a little clearer imo.)